### PR TITLE
[Merged by Bors] - feat: add `(Algebra|IntermediateField).lift_cardinalMk_adjoin_le`

### PR DESCRIPTION
**Pull request by acmepjz**
Also add `FreeAlgebra.cardinalMk_eq_max_lift`, etc. parallel to `MvPolynomial`.

Also fix the name `cardinal_lift_mk_le_max` -> `cardinalMk_le_max_lift` for `MvPolynomial`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

- [x] depends on: #19286

### CODE REVIEW

**Comment by acmepjz** [comment]
> Dependency changes
> 
> File	Base Count	Head Count	Change
> Mathlib.LinearAlgebra.FreeAlgebra	1179	1194	+15 (+1.27%)

Oops, seems that this result should moved to another file

File: Mathlib/FieldTheory/Cardinality.lean [hunk 1]
```diff
@@ -71,3 +72,25 @@ theorem Field.nonempty_iff {α : Type u} : Nonempty (Field α) ↔ IsPrimePow #
...
+
+theorem lift_cardinalMk_adjoin_le {E : Type v} [Field E] [Algebra F E] (s : Set E) :
+    Cardinal.lift.{u} #(adjoin F s) ≤ Cardinal.lift.{v} #F ⊔ Cardinal.lift.{u} #s ⊔ ℵ₀ := by
+  refine (Algebra.lift_cardinalMk_adjoin_le F s).trans_eq' (Cardinal.lift_inj.2 ?_)
+  let A := Algebra.adjoin F s
+  let K := FractionRing A
+  have hg : Function.Injective A.val := Subtype.val_injective
+  let f : K →ₐ[F] E := IsFractionRing.liftAlgHom hg
+  have hf : f.fieldRange = adjoin F s :=
+    IsFractionRing.liftAlgHom_fieldRange_eq_of_range_eq hg A.range_val
```
Comment (acmepjz) [inline Mathlib/FieldTheory/Cardinality.lean:88]: These 3 lines are very slow. Any suggestions to improve them?

File: Mathlib/FieldTheory/Cardinality.lean [hunk 2]
```diff
@@ -71,3 +72,25 @@ theorem Field.nonempty_iff {α : Type u} : Nonempty (Field α) ↔ IsPrimePow #
...
+  rw [← hf, show #f.fieldRange = #K from (AlgEquiv.ofInjectiveField f).toEquiv.symm.cardinal_eq,
+    FractionRing.cardinalMk]
```
Comment (alreadydone) [inline Mathlib/FieldTheory/Cardinality.lean:90]: Not sure what caused the particular slowness but here's a proof without using FractionRing which also gets rid of the extra import:
```suggestion
open Cardinal in
theorem lift_cardinalMk_adjoin_le {E : Type v} [Field E] [Algebra F E] (s : Set E) :
    Cardinal.lift.{u} #(adjoin F s) ≤ Cardinal.lift.{v} #F ⊔ Cardinal.lift.{u} #s ⊔ ℵ₀ := by
  rw [show ↥(adjoin F s) = (adjoin F s).toSubfield from rfl, adjoin_toSubfield]
  apply (Cardinal.lift_le.mpr (Subfield.cardinalMk_closure_le_max _)).trans
  rw [lift_max, sup_le_iff, lift_aleph0]
  refine ⟨(Cardinal.lift_le.mpr ((mk_union_le _ _).trans <| add_le_max _ _)).trans ?_, le_sup_right⟩
  simp_rw [lift_max, lift_aleph0, sup_assoc]
  exact sup_le_sup_right mk_range_le_lift _
```

**Review by alreadydone — APPROVED** [review]

File: Mathlib/FieldTheory/Cardinality.lean (hunk 2 above)
Comment (acmepjz) [reply Mathlib/FieldTheory/Cardinality.lean:90]: Thank you very much!

**Comment by acmepjz** [comment]
After the proof changes, `#find_home IntermediateField.lift_cardinalMk_adjoin_le` suggests that

```text
[Mathlib.FieldTheory.NormalClosure, Mathlib.RingTheory.Nullstellensatz, Mathlib.FieldTheory.Fixed, Mathlib.RingTheory.WittVector.FrobeniusFractionField, Mathlib.LinearAlgebra.Eigenspace.Triangularizable, Mathlib.RepresentationTheory.FDRep]
```

But none of them seems appropriate. Perhaps `Mathlib.FieldTheory.Cardinality` is OK, or is it too late?

**Comment by alreadydone** [comment]
Importing Mathlib.SetTheory.Cardinal.Subfield into FieldTheory.Adjoin only adds 3 trans imports (1261 to 1264), which is reasonable to me. Mathlib.FieldTheory.Cardinality imports 1307 files so indeed seems a bit "late".

**Comment by acmepjz** [comment]
> Importing Mathlib.SetTheory.Cardinal.Subfield into FieldTheory.Adjoin only adds 3 trans imports (1261 to 1264), which is reasonable to me. Mathlib.FieldTheory.Cardinality imports 1307 files so indeed seems a bit "late".

Done.

**Comment by alreadydone** [comment]
Thanks 🎉 
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**